### PR TITLE
Label updates

### LINKED ISSUE #3538: [FEATURE] Geometry generators for labeling
Original commit: https://github.com/qgis/QGIS/commit/857fe53773e5cb046953d88194ae5203e7617430 by web-flow

➿ Geometry generators for labeling ➿

### LINKED ISSUE #3779: [needs-docs] Make label map tools follow click-click behavior
Original commit: https://github.com/qgis/QGIS/commit/8b83a461f9131cdf971bde9ff19837ac247de15c by nyalldawson

It's rather jarring that the move label and rotate label tools
don't follow the rest of the application in using click-click
to move and rotate and rotate labels, and instead use the older
click-and-drag behaviour. This commit reworks the tools to use
click to start, click to end behaviour instead.

It also makes them respect the same conventions as the move
and rotate features tools, where a right click cancels the
move/rotate, and same with pressing "esc" mid-operation.

### LINKED ISSUE #4455: [FEATURE][labels] allow to delete custom label position
Original commit: https://github.com/qgis/QGIS/commit/99bf9687aa0feeeeaf9b1308f5707da5e17eaa06 by 3nids

by pressing DEL key when moving the label

### LINKED ISSUE #4456: [FEATURE][labels] allow to delete custom label rotation
Original commit: https://github.com/qgis/QGIS/commit/0e110cf01a29ebea9cf40f0166e0150f55f471ac by 3nids

by pressing DEL key when moving the label

### LINKED ISSUE #5481: [FEATURE] Allow use of color related HTML font tags in label text
Original commit: https://github.com/qgis/QGIS/commit/d4fbb6c3d632445097edf98bf2987ca20f64dd79 by nyalldawson

When enabled, this option will treat label contents as HTML, and
any FOREGROUND COLOR RELATED html formatting options will be
respected in the rendered label.

I repeat: ONLY HTML COLOR TAGS ARE RESPECTED. This is NOT a bug,
rather this feature has been designed as a "test of the waters"
with HTML formatting in labels, and accordingly only formatting
options which do not alter the font shape were considered.

On the plus side, it works correctly with ALL other label settings,
including shadows, buffers, curved labels, etc!

Sponsored by geoProRegio AG (www.geoproregio.ch)

### LINKED ISSUE #6560: Add a shortcut action to toggle labels in the vector layer right click menu (Request in QGIS)
### Request for documentation
From pull request QGIS/qgis#41935
Author: @nyalldawson
QGIS version: 3.20

**Add a shortcut action to toggle labels in the vector layer right click menu**

### PR Description:
Allows for labels to be quickly switched on or off, without losing the label configuration. If a layer has never had labeling configured and the action is checked then a default simple labeling is added to the layer.

Refs Natural resources Canada Contract: 3000720707

Peek 2021-03-01 16-05

### Commits tagged with [need-docs] or [FEATURE]
"[feature] Add a shortcut action to toggle labels in the vector layer right\nclick menu\n\nAllows for labels to be quickly switched on or off, without\nlosing the label configuration. If a layer has never had labeling\nconfigured and the action is checked then a default simple\nlabeling is added to the layer.\n\nRefs Natural resources Canada Contract: 3000720707"

### LINKED ISSUE #6591: When interactively moving a callout line, holding shift will cause the callout to snap to 15 degree angle increments (Request in QGIS)
### Request for documentation
From pull request QGIS/qgis#42217
Author: @nyalldawson
QGIS version: 

**When interactively moving a callout line, holding shift will cause the callout to snap to 15 degree angle increments**

### Commits tagged with [need-docs] or [FEATURE]
"[feature] When interactively moving a callout line, holding shift\nwill cause the callouti line angle to snap to 15 degree increments\n\nThis allows for easier creation of nicely parallel callout lines"

### LINKED ISSUE #6610: New callout style for curved lines (Request in QGIS)
### Request for documentation
From pull request QGIS/qgis#42302
Author: @nyalldawson
QGIS version: 3.20

**New callout style for curved lines**

### PR Description:
Renders a nice cartographically pleasing curved line between the
labels and features.

Options include selecting a specific curve orientation (clockwise
or counterclockwise), or an automatic orientation option which
determines optimal orientation for each individual label. Users
also have control over the amount of curvature applied to the
callout lines.

Here's a quick screencast of the automatic orientation option:

Peek 2021-03-18 16-04

### Commits tagged with [need-docs] or [FEATURE]
"[feature] New callout style for curved lines\n\nRenders a nice cartographically pleasing curved line between the\nlabels and features.\n\nOptions include selecting a specific curve orientation (clockwise\nor counterclockwise), or an automatic orientation option which\ndetermines optimal orientation for each individual label. Users\nalso have control over the amount of curvature applied to the\ncallout lines."

### LINKED ISSUE #6611: Add new "balloon" (speech bubble) callout style (Request in QGIS)
### Request for documentation
From pull request QGIS/qgis#42338
Author: @nyalldawson
QGIS version: 3.20

**Add new "balloon" (speech bubble) callout style**

### PR Description:
Adds a new balloon callout style, which uses the old qgis annotation appearance (speech bubble) to expose a callout style similar in appearance to ArcMap's "balloon" callouts:

image

Sponsored by SLYR

### Commits tagged with [need-docs] or [FEATURE]
"[feature] Add new \"balloon\" (speech bubble) callout style"

### LINKED ISSUE #6631: Add support for rounded corners for balloon callout styles (Request in QGIS)
### Request for documentation
From pull request QGIS/qgis#42650
Author: @nyalldawson
QGIS version: 

**Add support for rounded corners for balloon callout styles**

### PR Description:
Allows for visually pleasing rounded corners by exposing a "corner radius" setting for balloon callouts.

image

### Commits tagged with [need-docs] or [FEATURE]
"[feature] Add support for rounded corners for balloon callout styles\n\nAllows for visually please rounded corners by exposing a \"corner radius\"\nsetting for balloon callouts."

### LINKED ISSUE #6641: Add an option to control how line features are clipped before their anchor points are calculated (Request in QGIS)
### Request for documentation
From pull request QGIS/qgis#42727
Author: @nyalldawson
QGIS version: 

**Add an option to control how line features are clipped before their anchor points are calculated**

### PR Description:
The previous behavior was to always use only the visible part of the line when calculating the anchor point for labels. Now,
users have the option to choose "Use Entire Line" when setting the label anchor point for lines, so that the entire line
feature is considered when calculating the point to place the label instead of just the visible part of the line.

image

Sponsored by Stadt Zürich

### Commits tagged with [need-docs] or [FEATURE]
"[feature][labelling] Add an option to control how line features are\nclipped before their anchor points are calculated\n\nThe previous behavior was to always use only the visible part\nof the line when calculating the anchor point for labels. Now,\nusers have the option to choose \"Use Entire Line\" when setting\nthe label anchor point for lines, so that the entire line\nfeature is considered when calculating the point to place\nthe label instead of just the visible part of the line.\n\nSponsored by Stadt Zürich"

### LINKED ISSUE #6678: [feature][callouts] Add blending mode settings (Request in QGIS)
### Request for documentation
From pull request QGIS/qgis#42750
Author: @nirvn
QGIS version: 3.20

**[feature][callouts] Add blending mode settings**

## Description

The capability to change blending mode for pretty much everything being rendered in QGIS is a great strength of ours. Let's have callouts benefit from it too:
image

### Commits tagged with [need-docs] or [FEATURE]
"[feature][callouts] Add blending mode settings"

### PR DESCRIPTION
**Pull request by roya0045**
Fixes #5481 Fixes #3538 Fixes #6560 Fixes #6591 Fixes #6641 Fixes #6631 Fixes #6611 Fixes #6610

I have also added some images, would like feeback to know if I should replace or keep them or not. I didn't overwrite the existing content to be safe.

Layout may not be perfect as I'm a bumblefack.

Edit:
Fixes #4456 Fixes #4455 Fixes #3779
Fixes #6678

### CODE REVIEW

**Comment by roya0045** [comment]
@DelazJ Thanks for the tags! They may this quicker as it only took a few hours to go through, document and test all the label issues.

**Comment by roya0045** [comment]
@SrNetoChan This is why labeling issue is useful. Anyone can just focus on what needs to be done and minimises fumbling about in the issues, qgis and the doc.

**Comment by SrNetoChan** [comment]
If it works, it works!!

You could have chosen a different subject than labels 😂

**Comment by roya0045** [comment]
@SrNetoChan Labels seem to be the ones that were the most done. As I can't label and am not too familiar with the documentation in details. I just picked the category that seemed the most complete at the moment.

Also Nyall and Mat did a lot of changes so I felt like doing labels would close out the most tickets.

**Comment by SrNetoChan** [comment]
> @SrNetoChan Labels seem to be the ones that were the most done. As I can't label and am not too familiar with the documentation in details. I just picked the category that seemed the most complete at the moment.
> 
> Also Nyall and Mat did a lot of changes so I felt like doing labels would close out the most tickets.

I was kidding. Good job.

I agree that tackling all the related issues in one go is preferable. Labeling in other hand can be time consuming for the few contributors we have.

**Comment by roya0045** [comment]
@SrNetoChan Labeling is restricted, so I can't do a rough pass when I have some free time.

I think having a rough first pass would already be good enough.

**Comment by SrNetoChan** [comment]
> @SrNetoChan Labeling is restricted, so I can't do a rough pass when I have some free time.
> 
> I think having a rough first pass would already be good enough.

You need to have more permissions.

@DelazJ do you know if we have any list for people with intermediate permissions? Like manage issues and add labels.

**Comment by DelazJ** [comment]
> Labeling in other hand can be time consuming for the few contributors we have.

👍 It's and I'm still unsure of the plus-value but let's do it if there's a hope of attracting people. I'll give it another time this w-e

> Labeling is restricted,

Reason why I tried to create a specific group with triage, labeling right yesterday but I failed (with no specific error message). Maybe someone with qgis organisation admin rights can give it a try (@elpaso @mbernasocchi @rduivenvoorde)?

**Comment by DelazJ** [comment]
Let me try another way to create the group

**Comment by DelazJ** [comment]
Ok. I can't create a group but I could give you the needed rights, Alexis. Should be good.

**Comment by roya0045** [comment]
@DelazJ Thanks! I'll do some when I have some down time.

**Review by DelazJ — COMMENTED** [review]
@roya0045 Sorry for the delay but here's my review. I hope remarks are coherent because I made it in multiple steps.

About screenshots, generally speaking: 
- provide the whole dialog (with header)
- try to enable/fill some more properties than the default

But in this particular case here the options can be visualized in the map canvas, I'd very much prefer we provide visual rendering of the options, more examples than just the (long) dialogs. See eg callout or placement figures. I think these could be more helpful than the dialogs. Opinions?

Nice work and thanks for the cleanup 👍

File: docs/user_manual/style_library/label_settings.rst [hunk 1]
```diff
@@ -605,6 +618,21 @@ Common placement settings
...
 
+Geometry Generator
+^^^^^^^^^^^^^^^^^^
+
+The :guilabel:'Geometry Generator' section allows a user to atler the label geometry
+by using expressions. This can be usefulto perform displacement of the geomtry dynamically
+or to convert it to another geometry.
+Some other options include:
+
+-Use the @map_scale variable to calculate distances / sizes be zoom level independent.
```
[the remark this answers is not in the document]
Comment (roya0045) [reply docs/user_manual/style_library/label_settings.rst:630]: I took the examples provided by Matthias, but something like 'make_point($x+15*@map_scale,$y+15*@map_scale)'

**Comment by roya0045** [comment]
I'll look into including the newest one.

**Review by DelazJ — COMMENTED** [review]
@roya0045 Sorry for the late feedback.
Could you revert the screenshots, please? I'll try to update the existing ones, with the same settings as others in the page (I don't remember whether settings are available in our sample projects already).

**Comment by DelazJ** [comment]
> I'll look into including the newest one.

May I suggest we address this in another PR, please?

File: docs/user_manual/style_library/label_settings.rst [hunk 2]
```diff
@@ -605,6 +620,35 @@ Common placement settings
...
+* Use the @map_scale variable to calculate distances / sizes be zoom level independent.
+* Combined with the curved placement mode, creates a circular label around a point feature::
+
+     exterior_ring(make_circle($geometry, 20))
+-Use a geometry which is saved in another field "label_position"
+-Use the generated geometry from the symbology also for labeling
+* Add a label at the start and the end of a line feature::
+
+    collect_geometries( start_point($geometry), end_point($geometry) )
+-Use a smoothed line to label a river
```
[the remark this answers is not in the document]
Comment (roya0045) [reply docs/user_manual/style_library/label_settings.rst:650]: It depend on the scale of the map, in a sense the zoom of the map, not the layout.

**Comment by roya0045** [comment]
> > I'll look into including the newest one.
> 
> May I suggest we address this in another PR, please?

Sure thing, I was just thinking that including it would reduce its chance of being left pending for extended periods of time.

File: docs/user_manual/style_library/label_settings.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (DelazJ) [reply docs/user_manual/style_library/label_settings.rst:650]: it **depends**, right? The text says "independent".

[the remark this answers is not in the document]
Comment (roya0045) [reply docs/user_manual/style_library/label_settings.rst:650]: Well you could use the variable to hav the offset always be x meters or use the variable to offset the auxiliary placement based on scale. I'll tweak it.

File: docs/user_manual/style_library/label_settings.rst [hunk 3]
```diff
@@ -417,6 +423,10 @@ tab and activate the |checkbox| :guilabel:`Draw callouts`. Then you can:
      :guilabel:`Top right`, :guilabel:`Left middle`, :guilabel:`Right middle`,
      :guilabel:`Bottom left`, :guilabel:`Bottom center` and :guilabel:`Bottom right`).
 
+Callouts can also be controlled manually by using the Movel Label tool in the Labelling Toolbar.
+The start and end point of each callout can be moved this way. The nodes should be highlighted
+when the mouse pointer is nearby. If needed the Shift Key can be held during the movement.
+This will snap the point in a way that the angle between the two point in increment of 15 degree.
```
[the remark this answers is not in the document]
Comment (roya0045) [reply docs/user_manual/style_library/label_settings.rst:431]: clarified slightly.

**Comment by roya0045** [comment]
@DelazJ images have been removed.

**Comment by DelazJ** [comment]
Here we go @roya0045 . And thanks for your patience.

**Comment by roya0045** [comment]
@DelazJ Thanks you for the help! No need to thank me and this was quite quick, no patiente was needed for this PR ;)

**Pull request merged**